src/Bicep.Core/TypeSystem/UnassignableType.cs

```diff
@@ -6,29 +6,34 @@
 
 namespace Bicep.Core.TypeSystem
 {
-    public class ErrorTypeSymbol : TypeSymbol
+    public class UnassignableTypeSymbol : TypeSymbol
     {
         private const string ErrorTypeName = "error";
         private readonly ImmutableArray<ErrorDiagnostic> errors;
 
-        public ErrorTypeSymbol(ErrorDiagnostic error)
+        public UnassignableTypeSymbol(string name)
+            : base(name)
+        {
+        }
+
+        public UnassignableTypeSymbol(ErrorDiagnostic error)
             : base(ErrorTypeName)
         {
             this.errors = ImmutableArray.Create<ErrorDiagnostic>(error);
         }
 
-        public ErrorTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
+        public UnassignableTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
             : base(ErrorTypeName)
         {
             this.errors = errors.ToImmutableArray();
         }
 
         public override IEnumerable<ErrorDiagnostic> GetDiagnostics()
         {
-            return this.errors;
+            return this.errors.IsDefaultOrEmpty ? ImmutableArray<ErrorDiagnostic>.Empty : this.errors;
```

Comment:
Can't you just return `this.errors`?

Reply:
UnassignableType can be a namespace, in this case, errors is uninitialized and I noticed exceptions when collecting errors on an uninitialized list.

Options here:

1. Do: this.errors.IsDefault ? ...
2. In a new the constructor that receives "name" initialize this.errors as an empty array

#2 is what I actually did, I updated the current overloaded constructor (one that received name) and added TypeKind, and in here I added a check to see if this.errors.IsDefault is true, if yes, I initialize it as an empty immutable array.